### PR TITLE
Add Prereconcile check to Managed Cluster and Agent Pool

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Implements the pre-reconciliation check to skip reconciliation of managed clusters and agent pools that are currently provisioning.

**Prerequisites**

- [x] PR #2683 

**Special notes for your reviewer**:

Check out the changes in the recordings to see the difference in behaviour.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 1]
```diff
@@ -120,3 +123,61 @@ func secretsToWrite(obj *containerservice.ManagedCluster, adminCreds string, use
...
+	var _ conversion.Hub = managedCluster
+
+	// If the cluster is in a state that will reject any PUT, then we should skip reconciliation
+	// as there's no point in even trying.
+	// This allows us to "play nice with others" and not use up request quota attempting to make changes when we
+	// already know those attempts will fail.
+	if provisioningState := managedCluster.Status.ProvisioningState; provisioningState != nil {
+		if blockingManagedClusterProvisioningStates.Contains(*provisioningState) {
+			return extensions.SkipReconcile(
+					fmt.Sprintf("Managed cluster is in provisioning state %q", *provisioningState)),
```
Comment (super-harsh) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:177]: Some details about the resource would be helpful here?

File: v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	"github.com/pkg/errors"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+var _ extensions.PreReconciliationChecker = &ManagedClustersAgentPoolExtension{}
+
+// If an agent pool has a provisioningState in this set, it will reject any attempt to PUT a new state out of hand;
+// so there's no point in even trying. This is true even if the PUT we're doing will have no effect on the state of the
+// cluster.
+var blockingManagedClustersAgentPoolProvisioningStates = set.Make(
```
Comment (super-harsh) [inline v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go:28]: consider making states lowercase and below we can use `toLower` while matching states for consistency?

File: v2/api/containerservice/customizations/managed_cluster_extensions.go (hunk 1 above)
Comment (theunrepentantgeek) [reply v2/api/containerservice/customizations/managed_cluster_extensions.go:177]: We don't need them here; the message is consumed twice - once via a logger that already includes the details of the resource itself, and once as the condition on the resource in the cluster. No ambiguity on what is being described.

**Review by super-harsh — APPROVED** [review]

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 3]
```diff
@@ -120,3 +124,69 @@ func secretsToWrite(obj *containerservice.ManagedCluster, adminCreds string, use
...
 	return collector.Values()
 }
+
+var _ extensions.PreReconciliationChecker = &ManagedClusterExtension{}
+
+// If a managed cluster has a provisioningState in this set, it will reject any attempt to PUT a new state out of hand;
+// so there's no point in even trying. This is true even if the PUT we're doing will have no effect on the state of the
+// cluster.
+// These are all listed lowercase, so we can do a case-insensitive match.
+var blockingManagedClusterProvisioningStates = set.Make(
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:134]: Same comment as the other PR - use an allow-list instead as it only needs 3 values then

File: v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+)
+
+var _ extensions.PreReconciliationChecker = &ManagedClustersAgentPoolExtension{}
+
+// If an agent pool has a provisioningState in this set, it will reject any attempt to PUT a new state out of hand;
+// so there's no point in even trying. This is true even if the PUT we're doing will have no effect on the state of the
+// cluster.
+// These are all listed lowercase, so we can do a case-insensitive match.
+var blockingManagedClustersAgentPoolProvisioningStates = set.Make(
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go:30]: Use an allow list instead

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 5]
```diff
@@ -120,3 +124,69 @@ func secretsToWrite(obj *containerservice.ManagedCluster, adminCreds string, use
...
+	// the hub type has been changed but this extension has not
+	var _ conversion.Hub = managedCluster
+
+	// If the cluster is in a state that will reject any PUT, then we should skip reconciliation
+	// as there's no point in even trying.
+	// This allows us to "play nice with others" and not use up request quota attempting to make changes when we
+	// already know those attempts will fail.
+	if provisioningState := managedCluster.Status.ProvisioningState; clusterProvisioningStateBlocksReconciliation(provisioningState) {
+		return extensions.BlockReconcile(
+				fmt.Sprintf("Managed cluster is in provisioning state %q", *provisioningState)),
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:179]: Do we need to protect against nil deref here?

File: v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		}
+	}
+
+	// If the agent pool is in a state that will reject any PUT, then we should skip reconciliation
+	// as there's no point in even trying.
+	// This allows us to "play nice with others" and not use up request quota attempting to make changes when we
+	// already know those attempts will fail.
+	if provisioningState := agentPool.Status.ProvisioningState; agentPoolProvisioningStateBlocksReconciliation(provisioningState) {
+		return extensions.BlockReconcile(
+				fmt.Sprintf("Managed cluster agent pool is in provisioning state %q", *provisioningState)),
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go:81]: Do we need to protect against nil refer here? (There are other places in this file this applies to as well)

**Review by matthchr — CHANGES_REQUESTED** [review]
LGTM -- but I think should move to allowlist instead of blocklist

File: v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go (hunk 4 above)
Comment (theunrepentantgeek) [reply v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go]: Done.

File: v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go (hunk 6 above)
Comment (theunrepentantgeek) [reply v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go]: In the same way as the other change, now saferer.

File: v2/api/containerservice/customizations/managed_cluster_extensions.go [hunk 7]
```diff
@@ -120,3 +124,59 @@ func secretsToWrite(obj *containerservice.ManagedCluster, adminCreds string, use
...
+
+// If a managed cluster has a provisioningState not in this set, it will reject any attempt to PUT a new state out of
+// hand; so there's no point in even trying. This is true even if the PUT we're doing will have no effect on the state
+// of the cluster.
+// These are all listed lowercase, so we can do a case-insensitive match.
+var nonBlockingManagedClusterProvisioningStates = set.Make(
+	"succeeded",
+	"failed",
+	"canceled",
+	"ready",
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_cluster_extensions.go:138]: FWIW I think it'd probably be safe to not have this here as AKS doesn't have the same ARM contract violation as SQL.

File: v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+// If an agent pool has a provisioningState not in this set, it will reject any attempt to PUT a new state out of
+// hand; so there's no point in even trying. This is true even if the PUT we're doing will have no effect on the state
+// of the agent pool.
+// These are all listed lowercase, so we can do a case-insensitive match.
+var nonBlockingManagedClustersAgentPoolProvisioningStates = set.Make(
+	"succeeded",
+	"failed",
+	"canceled",
+	"ready",
```
Comment (matthchr) [inline v2/api/containerservice/customizations/managed_clusters_agent_pool_extensions.go:34]: Same comment here

**Review by matthchr — APPROVED** [review]

**Pull request merged**